nautobot/extras/models/groups.py

```diff
@@ -431,3 +436,114 @@ def clean(self):
 
         # Validate `filter` dict
         self.clean_filter()
+
+    def process_group_filters(self, qs=None, group=None):
+        # Start with this group's base queryset
+        if qs is None:
+            # Clone the base queryset so we can modify it.
+            qs = self.get_queryset().clone()
+
+        if group is None:
+            group = self
+            print(f"Processing group {group}...")
+
+        # Enumerate the filters. Recursing into any groups of groups.
+        for membership in group.dynamic_group_memberships.all():
+            group = membership.group
+            print(f"Processing group {group}...")
+            child_memberships = group.dynamic_group_memberships.all()
+            if child_memberships.exists():
+                qs = self.process_group_filters(qs, group)
+
+            operator = membership.operator
+
+            fs = self.filterset_class(group.filter, self.get_queryset())
+            print(f"{group} -> {group.filter} -> {operator}")
+            if operator == "union":
+                qs = qs.union(fs.qs)
+            elif operator == "difference":
+                qs = qs.difference(fs.qs)
+            elif operator == "intersection":
+                qs = qs.intersection(fs.qs)
+
+        return qs
+
+
+class DynamicGroupFilterActionChoices(models.TextChoices):
+    UNION = "union", "Union"
+    INTERSECTION = "intersection", "Intersection"
+    DIFFERENCE = "difference", "Difference"
+
+
+class DynamicGroupMembership(BaseModel):
+    """Intermediate model for associating filters to groups."""
+
+    group = models.ForeignKey(
+        "extras.DynamicGroup",
+        on_delete=models.CASCADE,
+        related_name="+",
+    )
+    parent_group = models.ForeignKey(
+        "extras.DynamicGroup", on_delete=models.CASCADE, related_name="dynamic_group_memberships"
+    )
+    operator = models.CharField(choices=DynamicGroupFilterActionChoices.choices, max_length=12)
+    weight = models.PositiveSmallIntegerField()
+
+    def __str__(self):
+        return f"{self.group}: {self.operator} ({self.weight})"
+
+    class Meta:
+        constraints = [
+            models.UniqueConstraint(fields=["group", "parent_group"], name="group_to_filter_uniq"),
+        ]
+        ordering = ["parent_group", "weight", "group"]
+
+    def clean(self):
+        super().clean()
+
+        if all(
+            [
+                self.filter is not None,
+                self.group is not None,
+            ]
+        ):
+            raise ValidationError({"__all__": "Filter and group cannot both be set"})
+
+        if self.filter.content_type != self.parent_group.content_type:
+            raise ValidationError({"filter": "ContentType for filter and parent_group must match"})
```

Comment:
I must be missing something here - where is `DynamicGroupMembership.filter` defined?

Reply:
Yeah it's not there anymore. This needs to be revised. The other problem is that in the current pattern of using `my_group.groups.add(other_group, through_defaults={"operator": "union", weight=10}`, is that this `clean()` method is never called.

So to enforce `clean()` on association,  a `m2m_changed:pre_add` signal receiver will need to be implemented to explicitly call `full_clean()` on any new `DynamicGroupMembership` object attempting to be associated.